src/eos-updater-apply.c

```diff
@@ -73,41 +79,304 @@ get_test_osname (void)
+       * the error FLATPAK_ERROR_ONLY_PULLED */
+      if (!local_error)
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED, "Flatpak installation should not have succeeded!");
+          cleanup_undeployed_flatpaks_in_installation (installation, pending_flatpak_ref_actions, i);
+          return FALSE;
+        }
+
+      /* Something unexpected failed, return early now.
+       *
+       * We are not able to meaningfully clean up here - the refs will remain
+       * in the flatpak ostree repo for the next time we want to install them
+       * but there's nothing in the public API that will allow us to get rid of
+       * them. */
+      if (!g_error_matches (local_error, FLATPAK_ERROR, FLATPAK_ERROR_ONLY_PULLED))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          cleanup_undeployed_flatpaks_in_installation (installation, pending_flatpak_ref_actions, i);
+          return FALSE;
+        }
+    }
+
+  return TRUE;
+}
+
+static GPtrArray *
+prepare_flatpaks_to_deploy (OstreeRepo    *repo,
+                            const gchar   *update_id,
+                            GCancellable  *cancellable,
+                            GError       **error)
+{
+  g_autoptr(GHashTable) flatpak_ref_actions_this_commit_wants = NULL;
+  g_autoptr(GHashTable) flatpak_ref_action_progresses = NULL;
+  g_autoptr(GHashTable) relevant_flatpak_ref_actions = NULL;
+  g_autoptr(GPtrArray) flatpaks_to_deploy = NULL;
+
+  flatpak_ref_actions_this_commit_wants = flatpak_ref_actions_for_commit (repo,
+                                                                          update_id,
+                                                                          cancellable,
+                                                                          error);
+
+  if (!flatpak_ref_actions_this_commit_wants)
+    return NULL;
+
+  flatpak_ref_action_progresses =
+    eos_updater_util_flatpak_ref_action_application_progress_in_state_path (cancellable,
+                                                                            error);
+
+  if (!flatpak_ref_action_progresses)
+    return NULL;
+
+  /* Filter the flatpak ref actions for the ones which are actually relevant
+   * to this system and figure out which flatpaks need to be pulled from
+   * there */
+  relevant_flatpak_ref_actions = eos_updater_util_filter_for_new_flatpak_ref_actions (flatpak_ref_actions_this_commit_wants,
+                                                                                      flatpak_ref_action_progresses);
+
+  /* Convert the hash table into a single linear array of flatpaks to pull. The
+   * reason we need a linear array is that on failure, we need to roll back
+   * any flatpaks which were deployed */
+  flatpaks_to_deploy = eos_updater_util_flatten_flatpak_ref_actions_table (relevant_flatpak_ref_actions);
+
+  if (!pull_flatpaks (flatpaks_to_deploy, cancellable, error))
```

Comment:
Hmm, should we not be pulling the flatpaks as part of eos-updater’s fetch step, rather than its apply step?

Reply:
This is a very good point and something I should have thought of earlier when I was designing this. But ....

I think if we were to do the flatpak pulling into the fetch step, it could introduce a lot more complexity and cause this PR to go from a 4000 line diff to 6 or 7000 lines (with commensurate time taken up in review churn). The reason being that we still need to cleanup the right undeployed flatpaks in both the `fetch` and the `apply` error case and I could see that becoming a bunch of complicated state to manage.

Would you be happy with keeping the flatpak pulling in the apply step for now and looking into doing the pulling in the fetch step in a follow-up?